docs/lib/routers.md

```diff
@@ -0,0 +1,69 @@
+# Associative Resource Router
+
+django-ansible-base provides an `AssociationResourceRouter` which will auto-construct /associate and /disassociate endpoints for related ManyToMany fields for your models.
+
+To use this router simply do the following:
+```
+from ansible_base.lib.routers import AssociationResourceRouter
+
+router = AssociationResourceRouter()
+
+router.register(
+    r'users',
+    views.UserViewSet,
+    related_views={
+        'teams': (views.TeamViewSet, 'teams'),
+        'organizations': (views.OrganizationViewSet, 'organizations'),
+    },
+)
+
+urlpatterns = [
+    path('api/service/v1/', include(router.urls)),
+]
+```
+
+This would create an endpoint called users with `related` objects teams and organizations. You would also end up with associate/disassociate urls off the related endpoints like:
+  * `/api/service/v1/users/:id/teams/associate`
+  * `/api/service/v1/users/:id/teams/disassociate`
+  * `/api/service/v1/users/:id/organizations/associate`
+  * `/api/service/v1/users/:id/organizations/disassociate`
+
+
+## Breaking down related views
+
+Related views are expressed in the following format:
+
+```
+    '<entry in API related field>': (<ViewSet for relation>, '<related field name>')
+```
+
+NOTE: Often times the `<entry in API related field>` will be the same as `<related field name>` but this is not always the case.
```

Comment:
I thought this wasn't the plan. Here I think you're talking about building the URLs, right? Like `/api/v2/organizations/2/users/`, where "entry in API related field" is "users" here. But that's not the full story. If you view the organization detail view at `/api/v2/organizations/2/` then it should have

```python
"related": {
  "users": "/api/v2/organizations/2/users/"
}
```

We construct this related dictionary inside an instance method for a common abstract _model_. If you were to register a different related name like `/api/v2/organizations/2/member_users/`, then we may be able to coherently link to it, but it would be very difficult to change the key in the "related" dictionary to match. I'm also concerned about the mechanic of the internal view name.

For a complex example, consider "notification_templates_success". The related name _is_ non-trivial. But the link name, field name, and view suffix all seem consistent

https://github.com/ansible/awx/blob/519fd22bece0e33fa17e5aa6591fc28cf48f565b/awx/main/models/base.py#L381

https://github.com/ansible/awx/blob/519fd22bece0e33fa17e5aa6591fc28cf48f565b/awx/api/urls/project.py#L41

https://github.com/ansible/awx/blob/519fd22bece0e33fa17e5aa6591fc28cf48f565b/awx/api/serializers.py#L1342

So those are 3 things that all have correspondence. The related view name (the 4th naming) is "project_notification_templates_success_list", so it breaks down at that point.

My view is that we should start out declaring that all 4 of these things must have correspondence. Ping @relrod  If there is a need for something to break this correspondence, I would really like to press the issue and ask why.

Reply:
Internally the only thing `<entry in API related field>` is used for at this point is for constructing the URL. So `'foo': (FooViewSet, 'some_m2m')` would stick it under `.../foo/...`.

It's badly named "related_name" in the code, but it's not really that. It's "URL component".

The view names are (now, as of the work @AlanCoding did in #121 and related PRs) always generated using the field name - the second component of the 2-tuple.

> If you were to register a different related name like `/api/v2/organizations/2/member_users/`, then we may be able to coherently link to it, but it would be very difficult to change the key in the "related" dictionary to match.

Correct, the key under the related dict in the API response is generated by the field name, too.

I've previously thought about letting the router expose the data passed in, so that in `CommonModel#related_fields` we could pull out that data. And that's feasible (you'd need some way to get at the router instance from the model, but it can likely be done) if we want to support this particular case, but I'm not sure how critical it is.

> My view is that we should start out declaring that all 4 of these things must have correspondence.

I think they are? The URL component can be customized as I said, but that doesn't matter much in terms of actually getting at the route by some predetermined name. You can change the key from `users` to `people` and everything would continue working except the URL would change.

Where do we deviate from the pattern?